### PR TITLE
Implement customization for object serializers

### PR DESCRIPTION
**Pull request by justing-bq**
Classes and properties can be marked up with the IonSerializer to be used for serialization/deserialization.

```c#
[IonSerializer(typeof(MyIonCarSerializer))]
public class Car
{
}
```

The IonSerializer Attribute must specify a IonSerializer with a no-arguments public default constructor.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Amazon.Ion.ObjectMapper.Test/IonObjectSerializerTest.cs [hunk 1]
```diff
@@ -136,6 +136,24 @@ public void SerializesAndDeserializesCustomPropertyNames()
             Check(TestObjects.fmRadio);
         }
 
+        [TestMethod]
+        public void SerializesAndDeserializesWithCustomObjectSerializer()
+        {
+            var serializer = new IonSerializer();
```
Comment (kylepbit) [inline Amazon.Ion.ObjectMapper.Test/IonObjectSerializerTest.cs:142]: Can you add a test where an object that is annotated is a child of another object?

**Review by kylepbit — COMMENTED** [review]
+1

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs [hunk 2]
```diff
@@ -73,6 +82,15 @@ public override object Deserialize(IIonReader reader)
 
         public override void Serialize(IIonWriter writer, object item)
         {
+            var customObjectSerializer = this.GetCustomObjectSerializer();
```
Comment (kylepbit) [inline Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:85]: I feel like this becomes simpler by using the `TryCustomObjectSerialize` paradigm instead.

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs [hunk 3]
```diff
@@ -25,6 +25,15 @@ public IonObjectSerializer(IonSerializer ionSerializer, IonSerializationOptions
 
         public override object Deserialize(IIonReader reader)
         {
+            var customObjectSerializer = this.GetCustomObjectSerializer();
```
Comment (kylepbit) [inline Amazon.Ion.ObjectMapper/IonObjectSerializer.cs:28]: I feel like this becomes simpler by using the TryCustomObjectDeserialize paradigm instead.

File: Amazon.Ion.ObjectMapper.Test/IonObjectSerializerTest.cs (hunk 1 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper.Test/IonObjectSerializerTest.cs:142]: Done.

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs (hunk 2 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonObjectSerializer.cs]: Done.

File: Amazon.Ion.ObjectMapper/IonObjectSerializer.cs (hunk 3 above)
Comment (justing-bq) [reply Amazon.Ion.ObjectMapper/IonObjectSerializer.cs]: Done.

File: Amazon.Ion.ObjectMapper.Test/IonObjectSerializerTest.cs [hunk 4]
```diff
@@ -140,18 +140,34 @@ public void SerializesAndDeserializesCustomPropertyNames()
...
+            IIonStruct serialized = StreamToIonValue(stream);
+
+            Assert.IsTrue(serialized.ContainsField("dog"));
+            var serializedDog = serialized.GetField("dog");
+
+            // These field names come from the custom serializer.
+            AssertContainsFields(serializedDog, new string[] {"Given Name", "Male or Female", "Classification"});
+
+            var deserialized = serializer.Deserialize<DogOwner>(stream);
+            Assert.IsTrue(TestObjects.Rover.Equals(deserialized.Dog));
```
Comment (kylepbit) [inline Amazon.Ion.ObjectMapper.Test/IonObjectSerializerTest.cs:170]: Should also test `deserialized` for completeness.

**Review by kylepbit — COMMENTED** [review]
+1

File: Amazon.Ion.ObjectMapper.Test/IonObjectSerializerTest.cs [hunk 5]
```diff
@@ -141,6 +141,18 @@ public void SerializesAndDeserializesCustomPropertyNames()
...
+        [TestMethod]
+        public void SerializesAndDeserializesWithCustomObjectSerializer()
+        {
+            Check(TestObjects.Rover);
+        }
+        
+        [TestMethod]
+        public void SerializesAndDeserializesChildObjectWithCustomObjectSerializer()
+        {
+            Check(TestObjects.JohnDoe);
```
Comment (kylepbit) [inline Amazon.Ion.ObjectMapper.Test/IonObjectSerializerTest.cs:153]: Do we need both tests? I think this test would be sufficient, given that it's testing simple fields as well as complex ones?

Comment (justing-bq) [reply Amazon.Ion.ObjectMapper.Test/IonObjectSerializerTest.cs:147]: Removed redundant test.

**Review by HomayounA — COMMENTED** [review]
+1

**Comment by justing-bq** [comment]
Closing in favor of https://github.com/amzn/ion-object-mapper-dotnet/pull/35/.

**Pull request closed without merging**